### PR TITLE
Smp related changes in cos_kernel_api, sl, llbooter, capmgr and other libraries

### PR DESCRIPTION
**Pull request by phanikishoreg**
### Summary of this Pull Request (PR)

This PR includes changes for multi-core in `capmgr`, `sl`, `llbooter`, `sched` and other core parts at the user-level. This is intended for review feedback. There are some recent updates/fixes that are only in rump_cfe_integration branch, will PR that for both henry to see the changes in RK and for other core parts of Cos.

##### cos_kernel_api
* Most importantly, the update here is to have mutual-ex locks. This was done mainly to avoid race-conditions in captbl expand or page bump expands.

##### sl
* cross-core api and implementation for a scheduler. 
* Addition of `ck` submodule for use of ring-buffers.
* parent->child communication with use of a "shared" ringbuffer between a pair of (parent, child_x). When a child calls `schedinit_child`, parent updates its data structures with child threads with a capmgr api for getting child threads and then it creates a shared memory for use between the parent and the calling child and returns the shared memory cbuf id.

##### kernel
* introspect a rcvcap to get it's core id.
* ipi support for cross-core asnds from @yzcode 

##### capmgr
* per-core data-structures
* ability to create a "SINV" if a request for asnd is made for a rcvcap with cpuid != current cpu.

##### sinv_async (components/lib/sinv_async)
* Library for cross-core invocations and same core cross-component async invocations.

##### cos_time.h 
* Library to have time conversion API in sl => sl_usec2cyc, sl_cyc2usec to be out of sl as this is required by components that don't need to include sl for using this conversion API.

These are the main features/bugfixes in this PR. 
I think the intention of this PR is 2 fold. 
1. To review cos_kernel_api changes that use locks to avoid races. 
2. Review rest of the code for the usual PR feedback.

So review priority:
1. cos_kernel_api locks
2. sl changes for cross-core
3. llbooter/capmgr/sched changes for cross core
4. sl changes for parent-child comm.
5. sinv_async library for cross-core or cross-comp async comm.
6. kernel changes for ipi, capmgr changes for SINV interposition for cross-core comm.
7. rest of the code (including unit test directories for testing different functionalities in this PR)

### Intent for your PR

Choose one (Mandatory):

- [x] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [ ] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):
@gparmer 

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [x] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG
- [x] Naming adhere's to the SG
- [x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [ ] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [ ] I've made an attempt to remove all redundant code
- [x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [x] I've commented appropriately where code is tricky
- [ ] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):

-I remember testing all the changes I've added through most of the test scripts in platform/i386/runscripts. I don't remember the exacts however.

### CODE REVIEW

**Comment by phanikishoreg** [comment]
I've fixed the merge conflicts. Yes, I believe it is. 
I briefly checked the current development branch (rump_cfe_integration) history and there is nothing significant required for multicore working.

There is however IOAPIC stuff that I need to prepare to merge if that is required by the visiting scholar or anyone before the paper deadline.

**Comment by phanikishoreg** [comment]
I request all those who are going to use `multi-core` to spend some time reviewing this PR. This is a PR originally intended for feedback but had to be merged for use before it was reviewed.

It's a 5K line PR and I'm pretty sure I would have neglected something or made some stupid assumptions/mistakes either for single-core or multi-core one of those was having all of micro_booter tests commented out (made a PR #371 for that!)!!! 

@gparmer @ryuxin @hungry-foolish <who else?>- Please please review this when you all have time!

File: src/components/implementation/capmgr/naive/cap_info.c [hunk 1]
```diff
@@ -72,63 +89,170 @@ cap_info_comp_init(spdid_t spdid, captblcap_t captbl_cap, pgtblcap_t pgtbl_cap,
...
 }
 
-asndcap_t
-cap_aepkey_asnd_get(cos_aepkey_t key)
+cap_t
+cap_channelaep_asnd_get(cos_channelkey_t key, capid_t *cap)
 {
-	struct cos_compinfo    *cap_ci = cos_compinfo_get(cos_defcompinfo_curr_get());
-	struct cap_aepkey_info *ak     = cap_aepkey_get(key);
+	struct cos_compinfo        *cap_ci = cos_compinfo_get(cos_defcompinfo_curr_get());
```
Comment (gparmer) [inline src/components/implementation/capmgr/naive/cap_info.c]: any chance this could be `NULL`?

**Comment by gparmer** [comment]
I agree with @phanikishoreg.  This would be a good PR for broader review.  At least @WenyuanShao should look at it, if not @ryuxin and @hungry-foolish.

File: src/components/include/cos_time.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#include <cos_debug.h>
+#include <cos_types.h>
+#include <cos_kernel_api.h>
+
+static inline cycles_t
+time_cyc_per_usec(void)
+{
+	static int cycs = 0;
+
+	if (unlikely(cycs <= 0)) cycs = cos_hw_cycles_per_usec(BOOT_CAPTBL_SELF_INITHW_BASE);
```
Comment (gparmer) [inline src/components/include/cos_time.h]: Add a comment at the top explaining that the hw capability is required.

In the future, we likely want two *variants*, one that uses the hw cap, and one that makes an invocation to get this value (and caches the value in the library).

**Pull request merged**